docs/RFCS/20200414_scheduled_jobs.md

```diff
@@ -0,0 +1,383 @@
+the next runnable job, and attempts to execute it.
+
+Inside the transaction, we select the next eligible job to run.
+
+```sql
+SELECT 
+  C.job, C.schedule, C.exec_spec, C.user,
+  (SELECT count(*) 
+   FROM system.jobs J
+   WHERE J.sched_id = C.sched_id AND job_status NOT IN ('failed', 'succeeded')
+  ) AS num_running,
+FROM system.crontab C
+WHERE next_run < current_timestamp()
+```
+
+If `num_running > 0` and policy setting `skip_wait == false`, then we need to
+wait for the previous invocation to complete.  To do this, we simply advance
+the `next_run` time by some amount and commit transaction.
+
+Once job execution policy constraints have been satisfied, 
+we can compute the next time this job is supposed to run.
+If the schedule is set, then compute the next_run based on the schedule.
+If not, this is a one-off job, set `next_run` to `NULL` (and update
+`change_info` to indicate that the schedule completed).
+
+Finally, execute the statement and commit the transaction.
+
+We have multiple venues on improving scalability and performance in the future releases,
+driven by the customer needs (Cockroach Cloud, etc): 
+  * Nodes may pick a random runnable job (instead of the oldest), to reduce contention
+  * If polling becomes problematic (scale), it can be replaced by sharding the jobs to run
+  * We can add a coordinator job responsible for finding lots of work to do and 
+    spawning these jobs
+  * Loading some (or all) cron state in memory to avoid rescans
+
+#### Pause/Resume
+Pausing the job involves setting `next_run` to NULL (and updating
+change info appropriately, e.g. reason="paused by an operator")
+
+Un-pausing the job involves computing the `next_run` time based on the `schedule`
+(and recording the change info)
+
+#### One-Off Jobs
+It is possible to use scheduled jobs to execute a one-off job.  A one-off job is represented
+by having an entry in the `system.crontab` table that has a `next_run` set, but does not
+have a `schedule`.
+
+## Monitoring 
+
+Since scheduled jobs is responsible for running unattended jobs, it is import that
+we provide visibility into the health of those jobs.
+
+The full details of the monitoring integration are outside the scope
+of this document, however, at the very least, we will export the following
+metrics:
+   * number of paused cron jobs
+   * number of jobs started
+   
+## CLI / Admin UI
+
+Interaction with the cron job system will require CLI and UI support, as well as changes
+to the existing statements.
+
+The actual details of those changes are outside the scope of this RFC, but
+at the very least, we will need the support for the following:
+  * Create/Modify/Delete cron job
+  * Pause/Unpause cron job
+  * Show backup history created by particular schedule
+
+Most of the above can be done by regular SQL statements; however, 
+we think that we would need to have a UI with more polish, e.g:
+   * CREATE FULL CLUSTER BACKUP WITH weekly SCHEDULE
+
+## Future work
+
```

Comment:
I have skimmed the document and am not seeing anything that would cause problems with multi-tenancy, but as you go about fully designing/implementing this please make sure to set this all up in a way that allows this to be available to tenants (in the far future). See https://docs.google.com/document/d/1_QRXGSrqQYktLwyJquv7NMuFut77zRqrS4UXiLA00_w/edit?ts=5e5fcde3&skip_itp2_check=true&pli=1 for the lead doc on that initiative.
Perhaps worthwhile too to mention in this doc that it should be "straightforward" for tenants to schedule jobs with some extra work (that is out of scope initially).

As I said, from the doc this looks like it would "just work". Red flags would be a reliance on the KV layer internals (gossip, liveness, nodeID, etc).

Reply:
Yeah, I thought about this a bit -- my only real concern right now is that if dispatch is an entirely sql-level thing, and we say we'll never have multiple tenants in the same sql process, then we're implying that we'll _always_ have at least one sql process for every tenant (and can't do e.g. on-demand spinning up or whatever) since otherwise no one would be watching the schedule. But that seems like something we can tackle later, if/when we want to get to zero processes per idle tenant.